File: app/src/main/java/de/luki2811/dev/coronainzidenzen/InternetRequest.java
package de.luki2811.dev.coronainzidenzen;

import android.content.Context;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Scanner;

public class InternetRequest extends Thread {

    URL url;
    JSONObject jsonObject;
    Context context;
    String fileName;

    InternetRequest(URL url, Context context, String fileName){
        setUrl(url);
        setContext(context);
        setFileName(fileName);
    }

    public void setUrl(URL url) {
        this.url = url;
    }

    public void setContext(Context context) {
        this.context = context;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public void setJsonObject(JSONObject jsonObject) {
        this.jsonObject = jsonObject;
    }

    public JSONObject getJsonObject() {
        return jsonObject;
    }


    @Override
    public void run() {
        setJsonObject(getJSONfromURL(url));
        System.out.println("run(): " + jsonObject);
        Datein file = new Datein(fileName);
        file.writeInFile(jsonObject.toString(), context);
    }
    public static JSONObject getJSONfromURL(URL url) {
        JSONObject jsonObject = null;

        HttpURLConnection urlConnection = null;
        if (url == null){
            System.err.println("URL is null");
        }
        try {
            urlConnection = (HttpURLConnection) url.openConnection();
            InputStream in = new BufferedInputStream(urlConnection.getInputStream());
            Scanner scanner = new Scanner(in, StandardCharsets.UTF_8.name());
            jsonObject = new JSONObject(scanner.useDelimiter("//Z").next());
        } catch(IOException | JSONException e){
            e.printStackTrace();
        }
        catch(NullPointerException e){
            e.printStackTrace();
            System.err.println("URL konnte nicht gefunden werden");
            System.exit(-1);
        }
        finally {
            urlConnection.disconnect();
        }
        return jsonObject;
    }
}

File: app/src/main/java/de/luki2811/dev/inzidenzencheck/InternetRequest.kt
package de.luki2811.dev.inzidenzencheck

import android.content.Context
import android.widget.Toast
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedInputStream
import java.io.IOException
import java.net.URL
import java.net.URLConnection
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.system.exitProcess

class InternetRequest internal constructor(private var url: URL, private var context: Context, private var fileName: String) :
    Thread() {
    private var jsonObject: JSONObject? = null

    override fun run() {
        jsonObject = getJSONfromURL(url)
        val file = Datein(fileName)
        if (jsonObject == null){
            Toast.makeText(context, context.getString(R.string.error_cant_load_data), Toast.LENGTH_LONG).show()
            return
        }
        file.writeInFile(jsonObject.toString(), context)
    }

    companion object {
        fun getJSONfromURL(url: URL?): JSONObject? {
            var jsonObject: JSONObject? = null
            val urlConnection: URLConnection
            try {
                urlConnection = url!!.openConnection()
            } catch (e: IOException) {
                e.printStackTrace()
                return null
            }
            try {
                BufferedInputStream(urlConnection.getInputStream()).use { `in` ->
                    val scanner = Scanner(`in`, StandardCharsets.UTF_8.name())
                    jsonObject = JSONObject(scanner.useDelimiter("//Z").next())
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: JSONException) {
                e.printStackTrace()
            } catch (e: NullPointerException) {
                e.printStackTrace()
                exitProcess(-1)
            }
            return jsonObject
        }
    }
}